### PR TITLE
ext_proc logging is consuming too much memory.

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yanjunxiang-google** [comment]
For reviewers: 

ext_proc filter logging memory consumption is currently unbounded.
This may cause memory issues when body sending mode is streaming and a lot of gRPC calls are sent and received to the side stream.

**Comment by yanjunxiang-google** [comment]
@htuch @yanavlasov @tyxia @mpwarres @stevenzzzz

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.h:70]: I think `deque` is a better choice than list

It provides random access that list doesn't and it has O(1) add and remove at both ends like list. I think add/remove in middle is not needed as here was a vector anyways.
Also I think deque has better performance (like memory locality/ cache) as it allocates contiguous memory while the list is a node-based container

**Review by tyxia — COMMENTED** [review]
Start with a high level comment (Will look at implementation details later)

Actually do we have circular buffer utility in Envoy that we can use? I know there is one in Google.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:70]: Understand that. However which one is more memory efficient?

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.h:70]: Compared with list, I believe is deque. 

As I mentioned above list is node-based container. each node is separately allocated. So the traversal operations of list will chase pointers to a new memory that results in a cache miss in most cases.  Moreover, changing from `vector` to `list` disable the random access ability it originally has.

If we want to maximize performance we probably should look for customized circular buffer. From a quick search, i found `SliceDeque` or `QuicheCircularDeque`. I am not sure if we really need them here.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:70]: In that case, how about let's just used std::deque as it is already used in ext_proc filter here:
https://github.com/envoyproxy/envoy/blob/64e334de55386060e3124a0e9471e5b60804a4c0/source/extensions/filters/http/ext_proc/processor_state.h#L52

IIUC, QuicheCircularDeque is not a Envoy utility, and SliceDeque is but had not been used in Envoy.

**Comment by rshriram** [comment]
suggestion: rather than limiting the list size, is it possible to collapse log entries? Or better still, devise something else for body streaming scenario. It is unclear if latency, status or callback state are conveying meaningful info for body streaming. If we are going to do the ring buffer, one suggestion is to keep a separate buffer for the body stat alone. This way, large bodies will not overrun stat info about requests/responses.

**Comment by tyxia** [comment]
I think @rshriram has provided a good idea/direction. Depends on how long this log refactor will take, I think this can be folded into our current solution Or the long term solution. 
For current circular buffer solution to work: Besides right data structure, another important thing is buffer size `kMaxGrpcLogging` : not too tight so we don't loose too much data but not too loose that caused huge memory consumption. 
@stevenzzzz @mpwarres What's your take on this and buffer size?

**Comment by yanjunxiang-google** [comment]
> suggestion: rather than limiting the list size, is it possible to collapse log entries? Or better still, devise something else for body streaming scenario. It is unclear if latency, status or callback state are conveying meaningful info for body streaming. If we are going to do the ring buffer, one suggestion is to keep a separate buffer for the body stat alone. This way, large bodies will not overrun stat info about requests/responses.

It should be doable to keep three separate buffers for gRPC call stats, one for header, body, and trailer each,  is doable.  Please confirm this is the design choice @tyxia @stevenzzzz @mpwarres . Also, please comment what's the size for each buffer.

**Comment by yanjunxiang-google** [comment]
Based on an offline discussion with @mpwarres , I feel cap the gRPC call stats within a bound won't cause issues.  
This is what I am thinking:
1) I don't see a reason to keep separate gRPC stats for header or body since for gRPC, it's just payload sent to server.
2)  We can keep the current solution: one buffer for all.  But also add a keep incrementing gRPC call counter in the filter_. So, the call_status, duration_.,, the old stats may get pushed out, but this counters will record how many calls happened, and  what's the call_status, duration_, for the last 256 calls.

Does this make sense?

**Comment by yanjunxiang-google** [comment]
Some math about the stats capacity: 256

each gRPC call stats is 24 bytes and gRPC call stats are kept for both inbound and outbound direction.
So, per stream the call buffer is up to : 256 * 2 * 24 = 12k
For 1000 stream, this is 12k * 1000 = 12Mb.

**Comment by htuch** [comment]
To @rshriram point on collapsing, can we just maintain a histogram and counters for body events (which are the ones that seem to be the problem). @mpwarres would that would for your requirements?

**Comment by KBaichoo** [comment]
ping @mpwarres

**Comment by mpwarres** [comment]
> To @rshriram point on collapsing, can we just maintain a latency histograms and counters for body events (which are the ones that seem to be the problem). @mpwarres would that would for your requirements?

Yes, so long as it distinguishes between request and response. Also, in the case of a gRPC stream failure, which IIRC with the current `vector<ExtProcLoggingInfo::GrpcCall>` would be represented (somewhat artificially) as a non-OK `GrpcStatus` for the last `GrpcCall` in the vector, how would that be represented using the histogram?

One way to resolve the `GrpcStatus` issue would be to move it up a level to be a single value per `ExtProcLoggingInfo`, which would more accurately represent the scope of the status.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:64]: This probably solved the issue we saw. But it seems that we will always `pop one and push one` once we reach the the boundary of `kMaxGrpcLogging` , which is not ideal for performance.

So I am thinking should we think about a more aggressive removal like remove first N elements so that we don't need to pop_front every time once it reaches boundary.

**Comment by yanjunxiang-google** [comment]
@mpwarres @tyxia, 
I am still looking for a design which can satisfy the application needs at same time bound the memory. Please share if you can.
Also, is the short term solution, like bound it with 256 entries as in the current, an issue? If not, can we commit this then turn this filter into stable. Then have a more optimized solution later on?

**Comment by tyxia** [comment]
> @mpwarres @tyxia, I am still looking for a design which can satisfy the application needs at same time bound the memory. Please share if you can. Also, is the short term solution, like bound it with 256 entries as in the current, an issue? If not, can we commit this then turn this filter into stable. Then have a more optimized solution later on?

Sure! let's fix this issue and optimize the design later (e.g. recycling logic etc). 

256 seems fine to me but I am wondering if we can keep it smaller if we really don't care. Hope to have input from @mpwarres as consumer of these data.

Also, can we simply pop out like kMaxGrpcLogging/4 (I don't have theory behind it) when we reach to kMaxGrpcLogging limit?

**Comment by yanjunxiang-google** [comment]
> > @mpwarres @tyxia, I am still looking for a design which can satisfy the application needs at same time bound the memory. Please share if you can. Also, is the short term solution, like bound it with 256 entries as in the current, an issue? If not, can we commit this then turn this filter into stable. Then have a more optimized solution later on?
> 
> Sure! let's fix this issue and optimize the design later (e.g. recycling logic etc).
> 
> 256 seems fine to me but I am wondering if we can keep it smaller if we really don't care those data. Hope to have input from @mpwarres as consumer of these data. Also, can we simply pop out like kMaxGrpcLogging/4 (I don't have a theory behind this number) when we reach to kMaxGrpcLogging limit? (Those two questions are not blocking issues from me)


1) Sure, reduce it even smaller is better from memory consumption perspective. But it all depends on application needs.  
2) Purging 64 entries once reaching capacity is fine. But the only thing is that if it never refill back, we are losing more data than necessary.
3) A design consideration:  Can we do this: having two separated buffers. One for header and trailer, and one for data 
3.1) a fixed size array for header and trailer:
log_info_header_trailer[0] for request header gRPC call stats
log_info_header_trailer[1] for request trailer gRPC call stats
log_info_header_trailer[2] for response header gRPC call stats
log_info_header_trailer[3] for response trailer gRPC call stats

3.2) a deque with capacity 32 for request and response data
log_info_data

This probably is a breaking change. 
log_info[

**Comment by mpwarres** [comment]
> > @mpwarres @tyxia, I am still looking for a design which can satisfy the application needs at same time bound the memory. Please share if you can. Also, is the short term solution, like bound it with 256 entries as in the current, an issue? If not, can we commit this then turn this filter into stable. Then have a more optimized solution later on?
> 
> Sure! let's fix this issue and optimize the design later (e.g. recycling logic etc).
> 
> 256 seems fine to me but I am wondering if we can keep it smaller if we really don't care those data. Hope to have input from @mpwarres as consumer of these data. Also, can we simply pop out like kMaxGrpcLogging/4 (I don't have a theory behind this number) when we reach to kMaxGrpcLogging limit? (Those two questions are not blocking issues from me)

Re: application requirements, we'd like to have:
1. gRPC call count by event type ({request,response} x {headers,trailers}, body), where "call" = ProcessingRequest/ProcessingResponse pair.
2. Latency distribution (or exact info) by event type. For headers/trailers, exact might be simpler, since there should be at most one call for each.
3. Overall stream status, and if there is a failure (e.g. message timeout), the event type being processed when the failure occurred.

Purging `GrpcCall`s from the current vector past some limit would be ok so long as we don't lose info about event types and failure--e.g., body call records should not evict call info for request or response headers/trailers. The purging approach would lose some statistical information about entries that were dropped (as opposed to maintaining a distribution), but seeing as it's a rare case, I think we could live with that.

**Comment by yanavlasov** [comment]
Would this approach work?

1. Store entries for request and response headers
2. Store up to N entries for body. If body messages exceed N, no more new entries will be stored.
3. Store entries for request/response trailers.

/wait-any

**Comment by htuch** [comment]
I really think this shuld be a histogram for body content, there is no value in maintaining independent log entries.

**Comment by yanjunxiang-google** [comment]
So, based on the comments, and if we want to satisfy the requirements, also using histograms for body duration. This is what I am thinkink:
 
1) for headers, and trailers, we can continue use the current data structure, one entry for each gRPC call.
2) For body sending, for failed gRPC calls, we continue use the current data structure since here will be at most one entry for such calls.

3) For body sending, for successful gRPCs,  we can use Histogram for latency distribution. And we will need one histogram for each of these 6 combinations :  {request, response } * { BufferedBody, BufferedPartialBody,  StreamedBody}
 Here, do we need to differentiate the different body mode? i.e, should we just have two latency histograms, like request-body and response-body?

**Comment by yanjunxiang-google** [comment]
> 

This solution is simple. But it will lose information for body sending once it reaches N. So, not sure whether it is acceptable.

**Comment by yanjunxiang-google** [comment]
> Would this approach work?
> 
> 1. Store entries for request and response headers
> 2. Store up to N entries for body. If body messages exceed N, no more new entries will be stored.
> 3. Store entries for request/response trailers.
> 
> /wait-any

This solution is simple. But it will lose information for body sending once it reaches N. So, not sure whether it is acceptable.

**Comment by jmarantz** [comment]
Assigning Kevin as maintainer in the general space; feel free to wait for Tianyu's approval first.

/wait (for main merge)

**Comment by yanjunxiang-google** [comment]
> Assigning Kevin as maintainer in the general space; feel free to wait for Tianyu's approval first.
> 
> /wait (for main merge)

Oh, thanks for looking into this. We are redesign the ext_proc gRPC call stats. Let me convert this PR to draft mode for now.

**Comment by yanjunxiang-google** [comment]
> 

HI, @mpwarres , please take a look the latest change and let me know if you have comments.  I think the data structure meets all the requirements mentioned above.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (yanavlasov) [inline test/extensions/filters/http/ext_proc/filter_test.cc:1283]: You may want to isolate the common code to make these tests smaller.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (yanavlasov) [inline test/extensions/filters/http/ext_proc/filter_test.cc:1360]: It would be good test test failure at other stages to make sure the stats looks correct.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Comment by yanjunxiang-google** [comment]
> @yanjunxiang-google . The failing tests in CD/CI seems to be related to your time change in event/mock.cc. Please take a look. `uninteresting mock calls` warning in general can be silenced by wrapping your mock class into testing::NiceMock, but that seems to be the changes to all those failing tests.
> 
> For your need of different time in filter_test.cc, just brainstorming from a quick scan, I am wondering alternatively have your considered [test_time](https://github.com/envoyproxy/envoy/blob/1f67c88f26744e1f7c21c0e65798d77d69efeac8/test/test_common/test_time_system.h#L17) ? @yanavlasov probably has more insights here.

Thanks for the comments!
I pushed a commit to fix the uninterested mock function call issue reported by CI.

For time, TestTimeSystem had been tried. It works in the sense of it will return a real clock number.  However, since the consecutive call happens too fast in the test, it returns identical number at startGrpcCall and finishGrpcCall thus the duration is zero, which can not help the test purpose. Thus we have to using MockTimeSystem in our tests.  Also, since with MockTimeSystem , we can deliberately set the duration, that helps us to verify the results, like total, max, min durations.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/filter_test.cc:1373]: Done. More tests are added to test failures in req/resp headers/body/trailer.

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/filter_test.cc]: done

File: test/extensions/filters/http/ext_proc/filter_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/ext_proc/filter_test.cc:100]: please try SimulatedTimeSystem. I had attempted to deprecate MockTimeSystem though there were are few difficult-to-remove cases. THe doc for MockTimeSystem has a TODO to me to get rid of it :)

PIng me if you have questions but I think for the most part things should be easier. You will control advancement of time from the test thread.

Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/filter_test.cc]: Sure, done!

After changing to use SimulatedTimeSystem, we are calling

test_time_->timeSystem().setMonotonicTime(MonotonicTime(std::chrono::microseconds(x)));

before each filter_callbacks_->dispatcher().timeSource().monotonicTime() call to get the right latency:

https://github.com/envoyproxy/envoy/blob/76667dd9454101a6f9d38b672555e076706d190a/source/extensions/filters/http/ext_proc/processor_state.cc#L31

https://github.com/envoyproxy/envoy/blob/76667dd9454101a6f9d38b672555e076706d190a/source/extensions/filters/http/ext_proc/processor_state.cc#L41

**Review by yanavlasov — DISMISSED** [review]

**Comment by yanjunxiang-google** [comment]
Thanks @yanavlasov for approval!

Kind ping @mpwarres @tyxia

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.h:64]: QQ: so we removed `ProcessorState::CallbackState` from new`GrpcCall` struct?

This field has been used widely (internally). It also provides the value to determine filter status. Can we keep it? It is just a enum so it should not bloat memory usage I think

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:64]: CallBackState basically is to tell the stats is for headers, trailers, or body.  With the latest data structure design which keeps separate header_stats, trailer_stats, body_stats, there is no need to keep that any more.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:82]: Thanks for the comments!

This is due to header_stats may not be generated if processing mode for headers is "NONE". We can keep a null pointer there in such case. 

Same for the other two stats.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.h:64]: From a quick look at how this field is being used in our code base, I feel it is more than just header, buffer, trailer etc. It also tell filter status. Also, It will requires a large amount of code changes for clients

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.h:82]: I see.  It can be std::optional<GrpcCall>, which is also no heap allocation and CPU friendly.

It is totally up to you, either works for me.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:64]: Based on my reading, CallbackState is only used for that purpose (identify it's header/trailer/body)for the stats parsing in client side.  Please ping me directly if you need more explanation on this change.

Overall, the requirement for the gRPC call stats is captured in this comment: https://github.com/envoyproxy/envoy/pull/27812#issuecomment-1594940341
Also, there are some comments stating that for body, only total_latency, max_latency, min_latency, call counts and last_call_status is interested.  So, the stats is designed this way.  Client should refactor based on this as well.  

Here are some examples:
grpcCalls(INBOUND).header_stats_ is the header stats for request.
grpcCalls(OUTBOUND).header_stats_ is the header stats for response.

grpcCalls(INBOUND).body_stats_ is the body stats for request.
grpcCalls(OUTBOUND).body_stats_ is the body stats for response.

grpcCalls(INBOUND).trailer_stats_ is the trailer stats for request.
grpcCalls(OUTBOUND).trailer_stats_ is the trailer stats for response.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:82]: yeah, that's an option. Let me keep it as unique_ptr here as it is also widely used in Envoy.

**Review by tyxia — APPROVED** [review]
LGTM, great work! Thanks

**Comment by mpwarres** [comment]
> > 
> 
> HI, @mpwarres , please take a look the latest change and let me know if you have comments. I think the data structure meets all the requirements mentioned above.

I agree--LGTM. Thanks!

**Review by mpwarres — APPROVED** [review]

**Review by yanavlasov — DISMISSED** [review]

**Comment by yanavlasov** [comment]
@yanjunxiang-google Please resolve conflict and I will merge after that.

/wait

**Comment by yanjunxiang-google** [comment]
> 

Done! Thanks!

**Comment by yanjunxiang-google** [comment]
Kind ping!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
